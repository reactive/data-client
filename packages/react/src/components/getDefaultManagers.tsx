import type { DevToolsConfig, Manager } from '@data-client/core';

import {
  NetworkManager,
  SubscriptionManager,
  PollingSubscription,
  DevToolsManager,
} from '../managers/index.js';

/* istanbul ignore next */
/** Returns the default Managers used by DataProvider.
 *
 * @see https://dataclient.io/docs/api/getDefaultManagers
 */
let getDefaultManagers: ({
  devToolsManager,
  networkManager,
  subscriptionManager,
}?: GetManagersOptions) => Manager[] = ({
  networkManager,
  subscriptionManager = PollingSubscription,
} = {}) =>
  subscriptionManager === null ?
    [constructManager(NetworkManager, networkManager)]
  : [
      constructManager(NetworkManager, networkManager),
      constructManager(SubscriptionManager, subscriptionManager),
    ];
/* istanbul ignore else */
if (process.env.NODE_ENV !== 'production') {
  getDefaultManagers = ({
    devToolsManager,
    networkManager,
    subscriptionManager = PollingSubscription,
  }: GetManagersOptions = {}): Manager[] => {
    if (networkManager === null) {
      console.error('Disabling NetworkManager is not allowed.');
      networkManager = {};
    }
    const nm = constructManager(NetworkManager, networkManager);
    const managers: Manager[] = [nm];
    if (subscriptionManager !== null) {
      managers.push(constructManager(SubscriptionManager, subscriptionManager));
    }
    if (devToolsManager !== null) {
      managers.unshift(
        devToolsManager instanceof DevToolsManager ? devToolsManager : (
          new DevToolsManager(devToolsManager, nm.skipLogging.bind(nm))
        ),
      );
    }
    return managers;
  };
}
export { getDefaultManagers };

function constructManager<
  M extends { new (...args: any): Manager },
  O extends InstanceType<M> | ConstructorArgs<M>,
>(Mgr: M, optionOrInstance: O): InstanceType<M> {
  return optionOrInstance instanceof Mgr ?
      (optionOrInstance as any)
    : (new Mgr(optionOrInstance) as any);
}

export type GetManagersOptions = {
  devToolsManager?: DevToolsManager | DevToolsConfig | null;
  networkManager?: NetworkManager | ConstructorArgs<typeof NetworkManager>;
  subscriptionManager?:
    | SubscriptionManager
    | ConstructorArgs<typeof SubscriptionManager>
    | null;
};

export type ConstructorArgs<T extends { new (...args: any): any }> =
  T extends { new (options: infer O): any } ? O : never;
