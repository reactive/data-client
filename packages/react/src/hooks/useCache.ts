import type {
  EndpointInterface,
  DenormalizeNullable,
  Schema,
  FetchFunction,
  ResolveType,
  NI,
} from '@data-client/core';
import { ExpiryStatus } from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from '../hooks/useController.js';

/**
 * Read an Endpoint's response if it is ready.
 *
 * `useCache` guarantees referential equality globally.
 * @see https://dataclient.io/docs/api/useCache
 */
export default function useCache<
  E extends Pick<
    EndpointInterface<FetchFunction, Schema | undefined, undefined | boolean>,
    'key' | 'schema' | 'invalidIfStale'
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E['key']>] | readonly [null]
): E['schema'] extends undefined | null ?
  E extends (...args: any) => any ?
    ResolveType<E> | undefined
  : any
: DenormalizeNullable<E['schema']> {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { data, expiryStatus, expiresAt, countRef } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  /*********** This block is to ensure results are only filled when they would not suspend **************/
  // This computation reflects the behavior of useSuspense/useFetch
  // It only changes the value when expiry or params change.
  // This way, random unrelated re-renders don't cause the concept of expiry
  // to change
  const expired = useMemo(() => {
    return (Date.now() > expiresAt || forceFetch) && key;
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend/loading even if it is not fresh
  const loading = expiryStatus !== ExpiryStatus.Valid && expired;
  /****************************************************************************************************/

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return useMemo(() => {
    // if useSuspense() would suspend, don't include entities from cache
    if (loading) {
      if (!endpoint.schema) return undefined;
      // TODO: use getResponse() once it just returns data
      return controller.getResponseMeta(endpoint, ...args, {
        ...state,
        entities: {},
      }).data as any;
    }
    return data;
    // key substitutes args + endpoint
    // we only need cacheResults, as entities are not used in this case
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key, data, loading, cacheResults]);
  /*********************** end block *****************************/
}
