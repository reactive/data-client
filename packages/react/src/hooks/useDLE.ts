import type {
  Denormalize,
  DenormalizeNullable,
  ErrorTypes,
  EndpointInterface,
  FetchFunction,
  Schema,
  ResolveType,
} from '@data-client/core';
import { ExpiryStatus } from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

type SchemaReturn<S extends Schema | undefined> =
  | {
      data: Denormalize<S>;
      loading: false;
      error: undefined;
    }
  | { data: DenormalizeNullable<S>; loading: true; error: undefined }
  | { data: DenormalizeNullable<S>; loading: false; error: ErrorTypes };

type AsyncReturn<E> =
  | {
      data: E extends (...args: any) => any ? ResolveType<E> : any;
      loading: false;
      error: undefined;
    }
  | { data: undefined; loading: true; error: undefined }
  | { data: undefined; loading: false; error: ErrorTypes };

/**
 * Use async data with { data, loading, error } (DLE)
 * @see https://dataclient.io/docs/api/useDLE
 */
export default function useDLE<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? AsyncReturn<E>
: SchemaReturn<E['schema']>;

export default function useDLE<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): {
  data: E['schema'] extends undefined | null ? undefined
  : DenormalizeNullable<E['schema']>;
  loading: boolean;
  error: ErrorTypes | undefined;
};

export default function useDLE<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(endpoint: E, ...args: readonly [...Parameters<E>] | readonly [null]): any {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  // eslint-disable-next-line prefer-const
  let { data, expiryStatus, expiresAt, countRef } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    return controller.fetch(endpoint, ...(args as any)).catch(() => {});
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend/loading even if it is not fresh
  const loading = expiryStatus !== ExpiryStatus.Valid && !!maybePromise;

  data = useMemo(() => {
    // if useSuspense() would suspend, don't include entities from cache
    if (loading) {
      if (!endpoint.schema) return undefined;
      // TODO: use getResponse() once it just returns data
      return controller.getResponseMeta(endpoint, ...args, {
        ...state,
        entities: {},
      }).data as any;
    }
    return data;
    // key substitutes args + endpoint
    // we only need cacheResults, as entities are not used in this case
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key, data, loading, cacheResults]);

  const error = controller.getError(endpoint, ...args, state);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return {
    data,
    loading,
    error,
  };
}
