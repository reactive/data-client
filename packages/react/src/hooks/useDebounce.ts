import { useEffect, useState, useTransition } from 'react';

/**
 * Keeps value updated after delay time
 *
 * @see https://dataclient.io/docs/api/useDebounce
 * @param value Any immutable value
 * @param delay Time in miliseconds to wait til updating the value
 * @param updatable Whether to update at all
 * @example
 ```
 const [debouncedQuery, isPending] = useDebounce(query, 200);
 const list = useSuspense(getThings, { query: debouncedQuery });
 ```
 */
export default function useDebounce<T>(
  value: T,
  delay: number,
  updatable = true,
): [T, boolean] {
  const [debouncedValue, setDebouncedValue] = useState(value);
  const [isPending, startTransition] = useTran();

  useEffect(() => {
    if (!updatable) return;

    const handler = setTimeout(() => {
      startTransition(() => setDebouncedValue(value));
    }, delay);
    return () => {
      clearTimeout(handler);
    };
  }, [value, delay, updatable]);

  return [debouncedValue, isPending];
}

// compatibility with older react versions
const useTran =
  useTransition ?? /* istanbul ignore next */ (() => [false, identityRun]);
const identityRun = (fun: (...args: any) => any) => fun();
