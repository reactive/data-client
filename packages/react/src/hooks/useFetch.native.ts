import { DenormalizeNullable, ExpiryStatus, NI } from '@data-client/core';
import {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
} from '@data-client/core';
import { useMemo } from 'react';
import { InteractionManager } from 'react-native';

import useCacheState from './useCacheState.js';
import useController from './useController.js';
import useFocusEffect from './useFocusEffect.native.js';

/**
 * Request a resource if it is not in cache.
 * @see https://dataclient.io/docs/api/useFetch
 */
export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? ReturnType<E>
: Promise<Denormalize<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): E['schema'] extends undefined | null ? ReturnType<E> | undefined
: Promise<DenormalizeNullable<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): Promise<any> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { expiryStatus, expiresAt } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;
    // if args is [null], we won't get to this line
    return controller.fetch(endpoint, ...(args as Parameters<E>));
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  useFocusEffect(() => {
    // revalidating non-suspending data is low priority, so make sure it doesn't stutter animations
    const task = InteractionManager.runAfterInteractions(() => {
      if (Date.now() > expiresAt && key) {
        controller.fetch(endpoint, ...(args as Parameters<E>));
      }
    });

    return () => task.cancel();
  }, []);

  return maybePromise;
}
