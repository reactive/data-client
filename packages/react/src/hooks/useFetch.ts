import { DenormalizeNullable, ExpiryStatus, NI } from '@data-client/core';
import {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
} from '@data-client/core';
import { useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from '../hooks/useController.js';

/**
 * Fetch an Endpoint if it is not in cache or stale.
 * @see https://dataclient.io/docs/api/useFetch
 */
export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? ReturnType<E>
: Promise<Denormalize<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): E['schema'] extends undefined | null ? ReturnType<E> | undefined
: Promise<DenormalizeNullable<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): Promise<any> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { expiryStatus, expiresAt } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  return useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    // if args is [null], we won't get to this line
    return controller.fetch(endpoint, ...(args as Parameters<E>));
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);
}
