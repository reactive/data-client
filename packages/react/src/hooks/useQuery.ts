import type {
  DenormalizeNullable,
  NI,
  Queryable,
  SchemaArgs,
} from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

/**
 * Query the store.
 *
 * `useQuery` results are globally memoized.
 * @see https://dataclient.io/docs/api/useQuery
 */
export default function useQuery<S extends Queryable>(
  schema: S,
  ...args: NI<SchemaArgs<S>>
  // INVALID results in undefined regardless of DenormalizeNullable
): DenormalizeNullable<S> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = JSON.stringify(args);

  // even though controller.get() is memoized, its memoization is more complex than
  // this so we layer it up to improve rerenders
  const { data, countRef } = useMemo(() => {
    return controller.getQueryMeta(schema, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [state.indexes, state.entities, key]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return data;
}
