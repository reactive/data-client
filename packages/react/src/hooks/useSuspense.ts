import { ExpiryStatus } from '@data-client/core';
import type {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
  DenormalizeNullable,
  ResolveType,
} from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

/**
 * Ensure an endpoint is available.
 * Suspends until it is.
 *
 * `useSuspense` guarantees referential equality globally.
 * @see https://dataclient.io/docs/api/useSuspense
 * @throws {Promise} If data is not yet available.
 * @throws {NetworkError} If fetch fails.
 */
export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? ResolveType<E>
: Denormalize<E['schema']>;

export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): E['schema'] extends undefined | null ? ResolveType<E> | undefined
: DenormalizeNullable<E['schema']>;

export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(endpoint: E, ...args: readonly [...Parameters<E>] | readonly [null]): any {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { data, expiryStatus, expiresAt, countRef } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    return (
      controller
        // if args is [null], we won't get to this line
        .fetch(endpoint, ...(args as Parameters<E>))
        .catch(() => {})
    );
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend even if it is not fresh
  if (expiryStatus !== ExpiryStatus.Valid && maybePromise) {
    throw maybePromise;
  }

  const error = controller.getError(endpoint, ...args, state);

  if (error) throw error;

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return data;
}
