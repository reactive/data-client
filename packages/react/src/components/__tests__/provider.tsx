// eslint-env jest
import {
  NetworkManager,
  Manager,
  Middleware,
  Controller,
} from '@data-client/core';
import { act, render } from '@testing-library/react';
import { CoolerArticleResource } from '__tests__/new';
import nock from 'nock';
import React, { useContext, Suspense, StrictMode } from 'react';

import { ControllerContext, StateContext } from '../../context';
import { useController, useSuspense } from '../../hooks';
import { payload } from '../../test-fixtures';
import DataProvider from '../DataProvider';
import { getDefaultManagers } from '../getDefaultManagers';

describe('<DataProvider />', () => {
  let warnspy: jest.Spied<any>;
  let debugspy: jest.Spied<any>;
  beforeEach(() => {
    warnspy = jest.spyOn(global.console, 'warn').mockImplementation(() => {});
    debugspy = jest.spyOn(global.console, 'info').mockImplementation(() => {});
  });
  afterEach(() => {
    warnspy.mockRestore();
    debugspy.mockRestore();
  });

  beforeAll(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Access-Token',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload);
  });

  afterAll(() => {
    nock.cleanAll();
  });

  it('should warn users about missing Suspense', () => {
    const Component = () => {
      const article = useSuspense(CoolerArticleResource.get, { id: 5 });
      return <div>{article.title}</div>;
    };
    const tree = (
      <DataProvider>
        <Component />
      </DataProvider>
    );
    const { getByText } = render(tree);
    const msg = getByText('Uncaught Suspense.');
    expect(msg).toBeDefined();
    expect(warnspy).toHaveBeenCalled();
    expect(warnspy.mock.calls).toMatchSnapshot();
  });

  it('should not warn if Suspense is provided', () => {
    const Component = () => {
      const article = useSuspense(CoolerArticleResource.get, { id: 5 });
      return <div>{article.title}</div>;
    };
    const tree = (
      <StrictMode>
        <DataProvider>
          <Suspense fallback="loading">
            <Component />
          </Suspense>
        </DataProvider>
      </StrictMode>
    );
    const { getByText } = render(tree);
    const msg = getByText('loading');
    expect(msg).toBeDefined();
    expect(warnspy).not.toHaveBeenCalled();
  });

  it('should not change dispatch function on re-render', () => {
    let dispatch;
    let controller;
    let count = 0;
    function DispatchTester() {
      controller = useController();
      dispatch = useController().dispatch;
      count++;
      return null;
    }
    const chil = <DispatchTester />;
    const tree = <DataProvider>{chil}</DataProvider>;
    const { rerender } = render(tree);
    expect(dispatch).toBeDefined();
    let curDisp = dispatch;
    let curController = controller;
    rerender(tree);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    rerender(<DataProvider>{chil}</DataProvider>);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    const managers: any[] = [new NetworkManager()];
    rerender(<DataProvider managers={managers}>{chil}</DataProvider>);
    expect(count).toBe(1);
    curDisp = dispatch;
    curController = controller;
    rerender(<DataProvider managers={managers}>{chil}</DataProvider>);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    rerender(
      <ControllerContext.Provider
        value={new Controller({ dispatch: () => Promise.resolve() })}
      >
        {chil}
      </ControllerContext.Provider>,
    );
    expect(curController).not.toBe(controller);
    expect(count).toBe(2);
  });

  it('should change state', () => {
    jest.useFakeTimers({ now: 50 });
    let ctrl: Controller | undefined = undefined;
    let state;
    let count = 0;
    function ContextTester() {
      ctrl = useController();
      state = useContext(StateContext);
      count++;
      return null;
    }
    const chil = <ContextTester />;
    const tree = <DataProvider>{chil}</DataProvider>;
    render(tree);
    expect(ctrl).toBeDefined();
    expect(state).toBeDefined();
    act(() => {
      ctrl?.setResponse(
        CoolerArticleResource.get,
        { id: 5 },
        { id: 5, title: 'hi', content: 'more things here' },
      );
    });
    expect(count).toBe(2);
    expect(state).toMatchSnapshot();
    jest.useRealTimers();
  });

  it('should ignore dispatches after unmount', async () => {
    class InjectorManager implements Manager {
      declare controller: Controller;

      cleanup() {}

      middleware: Middleware = controller => {
        this.controller = controller;
        return next => async action => {
          await next(action);
        };
      };
    }
    const injector = new InjectorManager();
    const managers = [injector, ...getDefaultManagers()];
    let resolve: (r: any) => void = () => {};
    const endpoint = CoolerArticleResource.get.extend({
      fetch() {
        return new Promise(res => {
          resolve = res;
        });
      },
    });
    const Component = () => {
      const article = useSuspense(endpoint, { id: 5 });
      return <div>{article.title}</div>;
    };
    const tree = (
      <DataProvider managers={managers}>
        <Suspense fallback="loading">
          <Component />
        </Suspense>
      </DataProvider>
    );
    const { getByText, unmount } = render(tree);
    const msg = getByText('loading');
    expect(msg).toBeDefined();
    unmount();
    expect(debugspy).not.toHaveBeenCalled();
    await injector.controller.setResponse(
      endpoint,
      { id: 5 },
      { id: 5, title: 'hi' },
    );
    expect(debugspy).toHaveBeenCalled();
    expect(debugspy.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "Action dispatched after unmount. This will be ignored.",
      ]
    `);
  });
});
