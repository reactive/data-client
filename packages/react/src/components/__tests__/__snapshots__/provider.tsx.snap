// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataProvider /> should change state 1`] = `
{
  "endpoints": {
    "GET http://test.com/article-cooler/5": "5",
  },
  "entities": {
    "CoolerArticle": {
      "5": {
        "content": "more things here",
        "id": 5,
        "title": "hi",
      },
    },
  },
  "entitiesMeta": {
    "CoolerArticle": {
      "5": {
        "date": 50,
        "expiresAt": 60050,
        "fetchedAt": 50,
      },
    },
  },
  "indexes": {},
  "lastReset": 0,
  "meta": {
    "GET http://test.com/article-cooler/5": {
      "date": 50,
      "expiresAt": 60050,
      "fetchedAt": 50,
      "prevExpiresAt": undefined,
    },
  },
  "optimistic": [],
}
`;

exports[`<DataProvider /> should warn users about missing Suspense 1`] = `
[
  [
    "Uncaught suspense.
Make sure to add your own Suspense boundaries: https://dataclient.io/docs/getting-started/data-dependency#boundaries",
  ],
]
`;
